src/caffe/layers/detectnet_transform_layer.cpp

```diff
@@ -0,0 +1,929 @@
+    mixChannels(
+        to_channels.data(), 2,
+        &result, 1,
+        to_mix.data(), to_mix.size()/2
+    );
+
+    // convert back to BGR colorspace:
+    cvtColor(result, result, COLOR_HSV2BGR);
+  } else {
+    result = img;
+    saturation = 1.0;
+  }
+
+  return saturation;
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::visualize_bboxlist(
+    const Mat3v& img,
+    const Mat3v& img_aug,
+    const vector<BboxLabel >& bboxlist,
+    const vector<BboxLabel >& bboxlist_aug,
+    const Dtype* transformed_label,
+    const AugmentSelection& as
+) const {
+  // create buffer for visualization:
+  static const size_t offset = 2;
+  Mat3b vis_pane(Size(
+      max(img.cols, img_aug.cols) + offset * 2,
+      img.rows + img_aug.rows + offset * 3
+  ), 0);
+  Rect img_rect(offset, offset, img.cols, img.rows);
+  Rect aug_rect(offset, img.rows + offset*2, img_aug.cols, img_aug.rows);
+  {
+    Mat3b temp_image;
+
+    // place unaugmented image at top-left:
+    img.convertTo(temp_image, CV_8UC3);
+    temp_image.copyTo(vis_pane(img_rect));
+
+    // place augmented image below unaugmented image:
+    img_aug.convertTo(temp_image, CV_8UC3, UINT8_MAX, UINT8_MAX/2.0);
+    temp_image.copyTo(vis_pane(aug_rect));
+  }
+
+  // label the un-augmented ground-truth:
+  Mat3b img_vis(vis_pane, img_rect);
+  static const Scalar validClassColor(0, 255, 0);
+  static const Scalar ignoredClassColor(127, 127, 127);
+  foreach_(const BboxLabel& label, bboxlist) {
+    Point p1(label.bbox.tl().x, label.bbox.tl().y);
+    Point p2(label.bbox.br().x, label.bbox.br().y);
+    const Scalar& color = coverage_->validClass(size_t(label.classNumber))
+        ? validClassColor
+        : ignoredClassColor;
+    // rectangle for class:
+    rectangle(img_vis, p1, p2, color, 1, 8);
+    // class label:
+    std::stringstream ss;
+    ss << size_t(label.classNumber);
+    setLabel(img_vis, ss.str(), p1, 0.5);
+  }
+
+  // label augmented ground-truth by draw coverage label visualization
+  //  in-place:
+  Mat3b aug_vis(vis_pane, aug_rect);
+  coverage_->visualize(transformed_label, aug_vis);
+
+  // draw box around original and augmented image panes:
+  rectangle(vis_pane, img_rect.tl(), img_rect.br(), Scalar(255, 255, 255), 1);
+  rectangle(vis_pane, aug_rect.tl(), aug_rect.br(), Scalar(255, 255, 255), 1);
+
+  // draw text:
+  static size_t cnt = 0;
+  if (this->phase_ == TRAIN) {
+    std::stringstream ss;
+    ss  << "Aug"
+        << " crop=" << as.crop.x << "," << as.crop.y
+        << " scale=" << setprecision(2) << as.scale
+        << " rot=" << setprecision(2) << as.degree
+        << " flip=" << (as.flip ? "y" : "n")
+        << " sat=" << setprecision(2) << as.saturation
+        << " hue=" << setprecision(2) << as.hue_rotation;
+
+    std::string str_info = ss.str();
+    setLabel(vis_pane, str_info, Point(0, vis_pane.rows - 25));
+
+    // TODO: buffer overrun
```

Comment:
Please fix this. For instance with a `stringstream`, C++11 or a `boost::lexical_cast`.